include/dali/core/exec/tasking/scheduler.h

```diff
@@ -0,0 +1,300 @@
+ * `Wait`
+ * and in `Releasable::Release`.
+ *
+ * The scheduler implements deadlock mitigation by ensuring that objects are acquired only
+ * when all preconditions of a task can be met.
+ *
+ * Once a task is submitted to the scheduler, its shared pointer reference is increased and the
+ * caller doesn't need to maintain a copy of the task pointer.
+ *
+ * AddTask vs AddSilentTask
+ * AddTask produces a TaskFuture object. This object can be used to wait for the task and get its
+ * result. In case of tasks that do not produce final results, this future is not needed. In that
+ * case the overhead of creating a future object can be avoided by using AddSilentTask.
+ * The user can still Wait for tasks without a future object. The only difference is that
+ * the results of silent tasks are only accessible by the registered consumers of its outputs.
+ */
+class Scheduler {
+  struct TaskPriorityLess {
+    bool operator()(const SharedTask &a, const SharedTask &b) const {
+      return a->Priority() < b->Priority();
+    }
+  };
+
+ public:
+  /** Removes a ready task with the highest priorty or waits for one to appear or
+   *  for a shutdown notification.
+   */
+  SharedTask Pop() {
+    std::unique_lock lock(mtx_);
+    task_ready_.wait(lock, [&]() { return !ready_.empty() || shutdown_requested_; });
+    if (ready_.empty()) {
+      assert(shutdown_requested_);
+      return nullptr;
+    }
+    auto ret = std::move(ready_.top());
+    assert(ret->state == TaskState::Ready);
+    ready_.pop();
+    ret->state = TaskState::Running;
+    return ret;
+  }
+
+  /** Submits a task for execution
+   */
+  void AddSilentTask(SharedTask task) {
+    if (task->state != TaskState::New)
+      throw std::logic_error("A task can be submitted only once.");
+    AddTaskImpl(std::move(task));
+  }
+
+  /** Submits a task for execution and gets a Future which can be used to get the output value
+   */
+  [[nodiscard("Use AddSilentTask if the result is not needed")]] TaskFuture AddTask(
+      SharedTask task) {
+    if (task->state != TaskState::New)
+      throw std::logic_error("A task can be submitted only once.");
+    auto res = task->results_;
+    AddTaskImpl(task);
+    return {std::move(task), std::move(res)};
+  }
+
+  /** Notifies the scheduler that a Waitable's state has changed and tasks waiting for it
+   *  may become ready.
+   */
+  void DLL_PUBLIC Notify(Waitable *w);
+
+  /** Waits for a task to become complete.
+   */
+  void Wait(Task *task);
+
+  void Wait(const SharedTask &task) {
+    Wait(task.get());
+  }
+
+  /** Makes all Pop functions return with an error value */
+  void Shutdown() {
+    std::lock_guard g(mtx_);
+    shutdown_requested_ = true;
+    task_ready_.notify_all();
+  }
+
+ private:
+  /** Checks if the task is ready and moves it from pending_ to ready_ list */
+  bool DLL_PUBLIC CheckTaskReady(SharedTask &task) noexcept;
+
+  void AddTaskImpl(SharedTask task) {
+    if (task->state != TaskState::New)
```

Comment:
Maybe we can remove this check from `AddTask` and `AddSilentTask`?

Reply:
Will do; here we can have an `assert`.